### PR TITLE
Feature/141 read pre post sql

### PR DESCRIPTION
**Pull request by zzeekk**
### What changes are included in the pull request?
- implement pre/postReadSql
- rename pre/postSql -> pre/postWriteSql
- implement token substitution in sql's

### Why are the changes needed?
See #141

Note: just remarked that this also includes the changes of PR #143. Lets wait until this one is merged...

### CODE REVIEW

**Comment by pgruetter** [comment]
Doesn't work for me yet, will try to find out why. But to make sure, the following should work, correct?
```
mariaTarget {
    type = JdbcTableDataObject
    connectionId = mariaDbJdbc
    table = {
      name = target
    }
    preReadSql = "select * from dual"
  }
```
Even when executing a SQL that should definitely fail, the action is executed correctly. I can't see the log message .. preReadSql is being executed.

**Comment by pgruetter** [comment]
Doesn't work for me yet. With the following config:
```
  mariaSource {
    type = JdbcTableDataObject
    connectionId = mariaDbJdbc
    table = {
      name = source
    }
    preReadSql = "select * from dual"

  }
```

I get an exception
Caused by: scala.StringContext$InvalidEscapeException: invalid escape '\{' not one of [\b, \t, \n, \f, \r, \\, \", \'] at index 0 in "\{(.*?)\}". Use \\ for literal \.

Any idea why?

**Comment by zzeekk** [comment]
>     preReadSql = "select * from dual"

Hi @pgruetter,
"select * from dual" is not a valid statement in MariaDb, it results in java.sql.SQLException: No tables used. But "select 1 from dual" should be fine. This works for me with the following config:
`
connections {
  mariaDbJdbc {
    type = JdbcTableConnection
    driver = com.mysql.cj.jdbc.Driver
    url = "jdbc:mysql://localhost:3306"
    authMode = {
      type = BasicAuthMode
      userVariable = "CLEAR#root"
      passwordVariable = "CLEAR#elcaelca"
    }
  }
}
dataObjects {
  test-pre-sql-db {
    type = JdbcTableDataObject
    connectionId = mariaDbJdbc
    table = {
      db = test
      name = test
    }
    preReadSql = "select * from dual"
  }
  test-pre-sql-csv {
    type = CsvFileDataObject
    path = "~{id}.csv"
  }
}
actions {
  test-pre-sql {
    type = CopyAction
    inputId = test-pre-sql-db
    outputId = test-pre-sql-csv
    metadata {
      feed = test-pre-sql
    }
  }
}
`
Can you retry?

File: src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala [hunk 1]
```diff
@@ -31,9 +35,15 @@ import org.apache.spark.sql.{DataFrame, SaveMode, SparkSession}
...
  * @param id unique name of this data object
- * @param createSql DDL-statement to be executed in prepare phase
- * @param preSql SQL-statement to be executed before writing to table
- * @param postSql SQL-statement to be executed after writing to table
+ * @param createSql DDL-statement to be executed in prepare phase, using output jdbc connection
+ * @param preReadSql SQL-statement to be executed in exec phase before reading input table, using input jdbc connection.
+ *                   Use tokens with syntax %{<spark sql expression>} to substitute with values from [[DefaultExpressionData]].
+ * @param postReadSql SQL-statement to be executed in exec phase after reading input table and before action is finished, using input jdbc connection
+ *                   Use tokens with syntax %{<spark sql expression>} to substitute with values from [[DefaultExpressionData]].
+ * @param preWriteSql SQL-statement to be executed in exec phase before writing output table, using outputv connection
```
Comment (zzeekk) [inline src/main/scala/io/smartdatalake/workflow/dataobject/JdbcTableDataObject.scala:43]:
```suggestion
 * @param preWriteSql SQL-statement to be executed in exec phase before writing output table, using output jdbc connection
```

File: src/main/scala/io/smartdatalake/util/misc/SparkExpressionUtil.scala [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+ * Utils to misuse spark sql expressions as expression language to substitute tokens with values from case class instances.
+ */
+private[smartdatalake] object SparkExpressionUtil {
+
+  val tokenStartChar = "%"
+  val tokenExpressionRegex = (tokenStartChar + """\{(.*?)\}""").r
+
+  /**
+   * Substitutes all tokens in a string by the expression defined by the token evaluated against the give case class instance.
+   * Token syntax: "${<spark sql expression>}
```
Comment (zzeekk) [inline src/main/scala/io/smartdatalake/util/misc/SparkExpressionUtil.scala:42]:
```suggestion
   * Token syntax: "%{<spark sql expression>}
```

File: src/main/scala/io/smartdatalake/util/misc/SparkExpressionUtil.scala [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+/**
+ * Utils to misuse spark sql expressions as expression language to substitute tokens with values from case class instances.
+ */
+private[smartdatalake] object SparkExpressionUtil {
+
+  val tokenStartChar = "%"
+  val tokenExpressionRegex = (tokenStartChar + """\{(.*?)\}""").r
+
+  /**
+   * Substitutes all tokens in a string by the expression defined by the token evaluated against the give case class instance.
```
Comment (zzeekk) [inline src/main/scala/io/smartdatalake/util/misc/SparkExpressionUtil.scala:41]:
```suggestion
   * Substitutes all tokens in a string by the expression defined by the token evaluated against the given case class instance.
```

**Pull request merged**